Doc/library/asyncio-tutorial/async-functions.rst

```diff
@@ -0,0 +1,327 @@
+    ...     return 123
+    ...
+    >>> async def main():
+    ...     obj = f3()
+    ...     result = await obj
+    ...     print(result)
+    ...
+    >>> asyncio.run(main())
+    123
+
+In the code above, the value of ``obj`` is *not* ``123`` when
+coroutine function ``f3`` is evaluated. Instead, ``obj`` is a
+*coroutine* object, and it will only get executed when the
+``await`` keyword is used.  Of course, you don't have to write
+code like this where you first get the coroutine and then
+use ``await`` on the object; simply evaluate the
+coroutine function and use ``await`` all in the same line.
+
+An Aside: Similarity To Generator Functions
+-------------------------------------------
+
+This has nothing to do with asyncio, but you might be interested
+to see how this difference between a function and a
+coroutine function is quite similar to the difference between
+functions and generator functions:
+
+.. code-block:: python3
+
+    >>> def g():
+    ...     yield 123
+    ...
+    >>> inspect.isfunction(g)
+    True
+    >>> inspect.isgeneratorfunction(g)
+    True
+
+If a function uses the ``yield`` keyword anywhere inside the function
+body, that function becomes a *generator function*, very similar to
+how a function declared with ``async def`` becomes a
+*coroutine function*. And, completing the comparison, if you
+evaluate a generator function, a *generator* object is returned, similar
+to how a coroutine function, when evaluated, returns a coroutine
+object:
+
+.. code-block:: python3
+
+    >>> def g():
+    ...     yield 123
+    ...
+    >>> obj = g()
+    >>> type(obj)
+    <class 'generator'>
+    >>> inspect.isgenerator(obj)
+    True
+
+Again, this doesn't have anything to do with asyncio, but
+the loose similarity between generator functions and
+coroutine functions might give you a useful framework for understanding
+the new coroutine functions.
+
+Terminology For Async Generators
+--------------------------------
+
+The previous section was useful for giving you a basic framework
+for understanding how coroutines and generator have similar
+characteristics. Here, we show how we can also make asynchronous
+generator functions!  It sounds much more complicated than it
+really is, so let's jump directly to some examples:
+
+.. code-block:: python3
+
+    >>> import asyncio
+    >>> async def ag():
+    ...     yield 123
+    ...
+    >>> async def main():
+    ...     async for value in ag():
+    ...         print(value)
+    ...
+    >>> asyncio.run(main())
+    123
+
+If you pretend for a second that the word "async" is temporarily
+removed from the code above, the behaviour of the generator
+should look very familiar to you (assuming you already know how
+Python's generators work). The generator function yields out
+values and these values are obtained by iterating over the
+generator.
+
+The difference now is of course the presence of those "async"
+words. The code sample doesn't a good reason *why* an async
```

Comment:
Should it be "doesn't have a good reason" or "doesn't give a good reason" ?

Reply:
Yep, good catch